### ISSUE TITLE
Regression tests for training_config.

### ISSUE THREAD

**Issue by brendan-ai2** [allenai/allennlp#3649]
**Is your feature request related to a problem? Please describe.**
We currently don't have tests that cover training end-to-end with real (or near real) training configs. This means that we have broken config in our repo. For instance:

https://github.com/allenai/allennlp/blob/2e811b2796f77e9e796fc3a91f6618fdac68973b/training_config/bidirectional_language_model.jsonnet#L52 no longer works as of https://github.com/allenai/allennlp/pull/3597. The breaking change itself was intentional and well considered, but relying on auditing by hand to catch and fix these is not a great solution.

**Describe the solution you'd like**
At some point we had a system that did this for some number of our models. Unfortunately it really trained full models (IIRC), so it was rather slow and expensive. With a modest amount of parametrization for each training config we could likely produce a sniff test variant that would train very quickly. (I'm thinking two epochs with a limited number of batches per epoch. Success could be as simple as not crashing or, maybe, a drop in loss on the second epoch.) We could then run this on master commits like we do with the sniff tests.

**Describe alternatives you've considered**
1. Train full models. Rather slow. Attempted previously.
2. Unit tests only. Only indirectly helps catch breakage. What we do now.

**Additional context**
We need to consider how this fits in with the current project to break `allennlp` into sub-repos. One option would be for the sub-repos to include their training configs in their python packages (somewhat non-standard, but they're small, so this wouldn't be too onerous). Then `allennlp-hub` could have tests that run them.

**Comment by matt-gardner** [comment]
An easy alternative to full-on training would be to just validate the configuration files.  E.g., after #3627 is done, we could do this just by calling `TrainModel.from_params()` on the config file.  Hmm, we'd probably want to set the dataset readers to lazy before doing that.  But something like that should be a quick and easy test.

**Comment by brendan-ai2** [comment]
I like that idea a lot!

**Comment by DeNeutoy** [comment]
We could also add this into the main training loop (with some tweaks to the config, like removing the dataset reader and vocab creation), so we pre-crash if it is mis-configured.

**Comment by ZhaofengWu** [comment]
This will also help ensure that the demo models are always compatible with the latest master (assuming that their configs are all in `training_configs`). Otherwise different demo models might need to depend on different AllenNLP versions, creating a lot of overhead.

**Comment by dirkgr** [comment]
I recently added a parameter to the dataset readers that lets you limit the number of instances returned. So we could make a minimal modification to set epochs to 1 and only run on 100 instances (enough to fill at least two batches for most configs). This should work on all model configs in the same way.

Only issue is that some dataset readers will still read everything, and then only return the first 100 instances. In those cases we either live with it, or we fix it in the readers.
